### ISSUE TITLE
Export data as C code

### ISSUE THREAD

**Issue by MatrixM9** [NationalSecurityAgency/ghidra#2635]
**Is your feature request related to a problem? Please describe.**
Currently when using **Export Programm** data, such as global variables is not exported alongside the functions. Exporting this information as well would be very valuable for the user.

For example creating C code for arrays, structure initializations, simple value assignments etc.

**Describe the solution you'd like**
Add an option to **Export Programm** to also export the data as C code.

**Describe alternatives you've considered**
Writing a script to automate this or exporting the data by hand.

**Comment by loneicewolf** [comment]
I agree with @MatrixM9 , This is something I would like to see. It would make the job easier for many.

Have a great week both **@MatrixM9**   And **@NSA!**

//Will

**Comment by astrelsky** [comment]
An alternative would be to add a copy special for this.

Maybe this could be done via extension point so that users could add their own to format it in the language of their choice.

**Comment by pabx06** [comment]
exporting data/structure  that you have defined in the listing. would be nice.
use case: you want to export an array of 200 bytes used for crypto routines. for later use...

**Comment by astrelsky** [comment]
> exporting data/structure  that you have defined in the listing. would be nice.
> use case: you want to export an array of 200 bytes used for crypto routines. for later use...

You can copy as C code now. Have you tried it out?

**Comment by pabx06** [comment]
> > exporting data/structure  that you have defined in the listing. would be nice.
> > use case: you want to export an array of 200 bytes used for crypto routines. for later use...
> 
> You can copy as C code now. Have you tried it out?

![1616946682846646603554126274967](https://user-images.githubusercontent.com/4013370/112758358-4060b480-8fee-11eb-8597-ee22a08125f4.jpg)
I tried all the above 
However the array of struct i am trying to copy has components like strings (char *) and int. The strings doesn't get copied out.. primitive are 👌 but pointer pointee doesn't

**Comment by EuanKirkhope** [comment]
If you clone snapshot the listing view, press the edit listing button, then setup as below:

![exportvars](https://user-images.githubusercontent.com/12889375/137608122-91e9b5d5-a581-426b-ab7c-7ffcd6f0d919.png)

Unselect the edit listing, and you can then cut and paste the text into a text editor and it gets you 90% of the way there.  Doesn't work to well for large byte arrays due the dreaded ellipsis' (...) but is a good start

Also needs a script or manual clean up of empty padding bytes

**Comment by wothke** [comment]
> An alternative would be to add a copy special for this.
> 
> Maybe this could be done via extension point so that users could add their own to format it in the language of their choice.

It is just plain annoying that the existing "Export program C/C++" function does not export the global variables that the exported functions actually use. It is all the more frustrating after having spent all that time manually configuring the correct data types and even giving labels to those vars. Obviously it is a bug of the existing export C/C++ functionality that it completely ignores the global variables.

**Comment by IceTDrinker** [comment]
where should one look to try to add some form of export for these globals ?

**Comment by lorddevereux** [comment]
This is still an issue as of 11.2.

The decompiler obviously understands the structure enough to display the fields (once defined) so it can't be too hard to add code generation for this. Can anyone point me in the right direction?
